### PR TITLE
[SOFT-341] Power Select Rev 2

### PR DESCRIPTION
**Pull request by hewittmcg**
Finally got around to finishing the power_selection -> power_select changes.  [This confluence page](https://uwmidsun.atlassian.net/wiki/spaces/ELEC/pages/1763737727/Power+Select+FW+Rev+2) offers a general overview of some of the elements needed.

Brief summary of what this module (should) do:

- Periodically read voltages and currents from all three power sources, as well as temperature from aux/DCDC
- Broadcast these values
- Check for whether any of the readings represent a fault.  If so, turn off the LTC and broadcast the fault bitset under `SYSTEM_CAN_MESSAGE_POWER_SELECT_FAULT`
- Handle interrupts from the DCDC fault pin and perform same fault handling as above
- Handle `SYSTEM_CAN_MESSGAGE_POWER_ON_MAIN_SEQUENCE` messages by checking AUX or DCDC status
- Handle `SYSTEM_CAN_MESSGAGE_POWER_ON_AUX_SEQUENCE` messages by checking AUX status

### CODE REVIEW

**Comment by hewittmcg** [comment]
Realized I never made this comment, but I have no idea why CI is failing like this.  Merged this branch into a local branch off master and could build without issues.

File: projects/power_select/inc/power_select.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+#include "status.h"
+
+#include "power_select_defs.h"
+#include "power_select_events.h"
+#include "power_select_thermistor.h"
+
+#define POWER_SELECT_MEASUREMENT_INTERVAL_MS 1000
+#define POWER_SELECT_MEASUREMENT_INTERVAL_US ((POWER_SELECT_MEASUREMENT_INTERVAL_MS)*1000)
+
+typedef enum {
```
Comment (JarvisWeng) [inline projects/power_select/inc/power_select.h:24]: Do these enums and `#defines` need to be here? Or can they go into their own `.h`

File: projects/power_select/inc/power_select_defs.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#define POWER_SELECT_DCDC_VALID_ADDR \
+  \
+{                                 \
+    GPIO_PORT_A, 9                   \
+  }
+#define POWER_SELECT_AUX_VALID_ADDR \
+  \
+{                                \
+    GPIO_PORT_A, 8                  \
+  }
```
Comment (JarvisWeng) [inline projects/power_select/inc/power_select_defs.h:87]: Nit: can this be on 1 line or 2 like above

File: projects/power_select/src/power_select_can.c [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#include "power_select_can.h"
+
+static CanAckStatus prv_confirm_aux(uint16_t fault_bitset, uint8_t valid_bitset) {
+  if (!(valid_bitset & 1 << POWER_SELECT_AUX_VALID) ||
+      fault_bitset & 1 << POWER_SELECT_AUX_OVERCURRENT ||
+      fault_bitset & 1 << POWER_SELECT_AUX_OVERVOLTAGE) {
```
Comment (JarvisWeng) [inline projects/power_select/src/power_select_can.c:6]: Nit: Maybe put brackets around conditions just for clarity

File: projects/power_select/rules.mk [hunk 4]
```diff
@@ -6,6 +6,7 @@
 # $(T)_SRC: $(T)_DIR/src{/$(PLATFORM)}/*.{c,s}
 
 # Specify the libraries you want to include
-$(T)_DEPS := ms-helper ms-common codegen-tooling
+$(T)_DEPS := ms-common
 
-$(T)_test_power_selection_MOCKS := adc_read_raw
\ No newline at end of file
+$(T)_test_power_select_MOCKS := adc_read_converted_pin gpio_get_state gpio_set_state
```
Comment (JarvisWeng) [inline projects/power_select/rules.mk:11]: Sorry do you mock these?

**Comment by JarvisWeng** [comment]
> Realized I never made this comment, but I have no idea why CI is failing like this. Merged this branch into a local branch off master and could build without issues.

Perhaps you need to include `can_msg_defs.h ` or just make a fake can id. Since it's just a test

File: projects/power_select/rules.mk (hunk 4 above)
Comment (hewittmcg) [reply projects/power_select/rules.mk:11]: Yeah, lines 63/93/115 in `test_power_select.c`

**Comment by hewittmcg** [comment]
> > Realized I never made this comment, but I have no idea why CI is failing like this. Merged this branch into a local branch off master and could build without issues.
> 
> Perhaps you need to include `can_msg_defs.h ` or just make a fake can id. Since it's just a test

Yeah, I thought it would be okay since I think it's included through the `can_pack.h` include in `power_select_can.h` but maybe there's something weird that happens with the linking? Pushed code adding that include to see if it fixes the issue, but will change the ID next if that doesn't help

**Comment by copperium** [comment]
Oh man, I know why CI is failing: the CI script runs `make codegen` before building anything (to verify that can_messages.asciipb has correct syntax I guess), which overwrites libraries/codegen-tooling headers with the headers generated from codegen/can_messages.asciipb. So if you changed the codegen headers by copy/pasting from the old codegen-tooling-msxiv repo, the changes will get overwritten in CI!

I'll change the CI script to check for this issue, but make the changes in codegen in this repo to fix it.

**Comment by copperium** [comment]
Update: implemented the CI script changes, now I'm gonna show off the cool auto commenting feature

**Review by copperium — CHANGES_REQUESTED** [review]
Looks great, the only functionality I can see lacking is the "some sort of alert when the battery voltage drops below 2V" for HW rev 2 from https://uwmidsun.atlassian.net/l/c/pJvuu1v9. I don't think that hardware requirement existed when this project was started though lol, so that can easily be added in another ticket. Comments are mainly nitpicks, nothing too major.

File: projects/power_select/inc/power_select.h [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  [POWER_SELECT_AUX] = POWER_SELECT_AUX_MAX_VOLTAGE_MV,
+  [POWER_SELECT_DCDC] = POWER_SELECT_DCDC_MAX_VOLTAGE_MV,
+  [POWER_SELECT_PWR_SUP] = POWER_SELECT_PWR_SUP_MAX_VOLTAGE_MV,
+};
+
+static const uint16_t MAX_CURRENTS[NUM_POWER_SELECT_CURRENT_MEASUREMENTS] = {
+  [POWER_SELECT_AUX] = POWER_SELECT_AUX_MAX_CURRENT_MA,
+  [POWER_SELECT_DCDC] = POWER_SELECT_DCDC_MAX_CURRENT_MA,
+  [POWER_SELECT_PWR_SUP] = POWER_SELECT_PWR_SUP_MAX_CURRENT_MA,
+};
```
[the remark this answers is not in the document]
Comment (hewittmcg) [reply projects/power_select/inc/power_select.h]: Yeah, makes sense, thanks -- I went the `extern` route.  

The PD naming format is a loooooot cleaner looking than what I originally had 😄

**Comment by hewittmcg** [comment]
> Looks great, the only functionality I can see lacking is the "some sort of alert when the battery voltage drops below 2V" for HW rev 2 from https://uwmidsun.atlassian.net/l/c/pJvuu1v9. I don't think that hardware requirement existed when this project was started though lol, so that can easily be added in another ticket. Comments are mainly nitpicks, nothing too major.

Oh yeah, good catch, it OK for me to throw up a separate ticket for that if there isn't one already?

**Comment by copperium** [comment]
Yep sure, throw up a ticket for it

**Comment by hewittmcg** [comment]
> Yep sure, throw up a ticket for it

Great, done, it's SOFT-475

**Comment by copperium** [comment]
I think you may have missed the comments under the "7 hidden conversations" thing

**Comment by hewittmcg** [comment]
> I think you may have missed the comments under the "7 hidden conversations" thing

Whoops yeah I missed those, thanks for the heads-up.  Will have them up in a bit

File: projects/power_select/src/power_select.c [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+      if (s_storage.currents[i] > MAX_CURRENTS[i]) {
+        s_storage.fault_bitset |= 1 << fault;
+        prv_handle_fault();
+      } else {
+        s_storage.fault_bitset &= ~(1 << fault);
+      }
+    } else {
+      s_storage.currents[i] = 0;
+    }
+    LOG_DEBUG("Current %d: %d\n", (int)i, (int)s_storage.currents[i]);
```
[the remark this answers is not in the document]
Comment (hewittmcg) [reply projects/power_select/src/power_select.c]: I couldn't figure out a way to abstract this away easily, not sure if I'm just blanking, though.

If similar code was repeated a couple additional times I'd implement it with a struct or something to store all the data specific to each iteration (scaling factor, pin address, which current to use, etc.) but I think that'd be overkill here.

I did factor out the internal workings of `prv_periodic_measure` into smaller helper functions to make its expected behaviour easier to understand at first glance, however. 

Let me know whether this makes sense to you

**Review by copperium — CHANGES_REQUESTED** [review]
Only thing blocking is the question about disabling the LTC, rest are small nitpicks.

File: projects/power_select/src/power_select.c [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+  [POWER_SELECT_AUX] = POWER_SELECT_AUX_MAX_CURRENT_MA,
+  [POWER_SELECT_DCDC] = POWER_SELECT_DCDC_MAX_CURRENT_MA,
+  [POWER_SELECT_PWR_SUP] = POWER_SELECT_PWR_SUP_MAX_CURRENT_MA,
+};
+
+// Broadcast the fault bitset and turn off/on the LTC as required
+static void prv_handle_fault(void) {
+  // Turn off/on LTC depending on fault status
+  GpioAddress pin = POWER_SELECT_LTC_SHDN_ADDR;
+  if (s_storage.fault_bitset == 0) {
```
[the remark this answers is not in the document]
Comment (hewittmcg) [reply projects/power_select/src/power_select.c:45]: Yeah makes sense, I'll ping hardware about it

**Review by ryandancy — APPROVED** [review]

**Pull request merged**